logserver/src/main/java/alluxio/logserver/AlluxioLogServerProcess.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.logserver;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.retry.ExponentialBackoffRetry;
+import alluxio.retry.RetryPolicy;
+import alluxio.util.CommonUtils;
+import alluxio.util.WaitForOptions;
+
+import org.apache.log4j.Hierarchy;
+import org.apache.log4j.Level;
+import org.apache.log4j.PropertyConfigurator;
+import org.apache.log4j.spi.LoggerRepository;
+import org.apache.log4j.spi.RootLogger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Function;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.Properties;
+
+/**
+ * A centralized log server for Alluxio
+ *
+ * Alluxio masters and workers generate logs and store the logs in local storage.
+ * {@link AlluxioLogServerProcess} allows masters and workers to "push" their logs to a
+ * centralized log server where another copy of the logs will be stored.
+ */
+public class AlluxioLogServerProcess implements LogServerProcess {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioLogServer.class);
+  private static final long STOP_TIMEOUT_IN_MS = 60000;
+  private static final int BASE_SLEEP_TIME_IN_MS = 50;
+  private static final int MAX_SLEEP_TIME_IN_MS = 30000;
+  private static final int MAX_NUM_RETRY = 20;
+
+  private final String mBaseLogsDir;
+  private int mPort;
+  private ServerSocket mServerSocket;
+  private final int mMinNumberOfThreads;
+  private final int mMaxNumberOfThreads;
+  private ExecutorService mThreadPool;
+  private volatile boolean mStopped;
+
+  /**
+   * Construct an {@link AlluxioLogServerProcess} instance.
+   *
+   * @param baseLogsDir base directory to store the logs pushed from remote Alluxio servers
+   */
+  public AlluxioLogServerProcess(String baseLogsDir) {
+    mPort = Configuration.getInt(PropertyKey.LOG_SERVER_PORT);
+    mMinNumberOfThreads = Configuration.getInt(PropertyKey.MASTER_WORKER_THREADS_MIN) + 2;
+    mMaxNumberOfThreads = Configuration.getInt(PropertyKey.MASTER_WORKER_THREADS_MAX) + 2;
```

Comment:
why do we limit the number of threads?

Reply:
Because we know there is an upper limit on the number of threads. It is bounded by the number of Alluxio servers in the cluster.

Reply:
ack